README.md
## LEFTOVERS AGAIN: A Pokemon AI Battle Arena

Live server running here: http://cyberdyne.psim.us

'Getting Started' guide [here](https://github.com/dramamine/leftovers-again/blob/master/FAQ.md).
Full code documentation [here](http://metal-heart.org/static/docs/).

This is the official repository for an ongoing AI competition for Pokemon battles.

Coding your own Pokemon AI bot doesn't have to be a major, time-consuming project. You can have your own bot up and running with <100 lines of code!

1997: Deep Blue defeats Kasparov at chess. 2016: AlphaGo annihilates the world's greatest Go players. 20XX: AI overtakes human players at the greatest game of all, Pokemon.

## Designed with Simplicity and Accessibility in Mind

Many developers have worked on AI bots before, in multiple languages, always having to start from the beginning - figuring out the Showdown protocol, importing relevant data, translating teams from human-readable text to code.

That work is *already done for you*.

You can have a working bot up and running in minutes. Read the [Getting Started guide](https://github.com/dramamine/leftovers-again/blob/master/FAQ.md) to see how easy it is.

![generate](https://user-images.githubusercontent.com/1554498/29701153-de0017b0-891e-11e7-80fc-60c7f7157a7e.gif)

_generating your first bot_

![run bot](https://user-images.githubusercontent.com/1554498/29701154-de093c0a-891e-11e7-9251-80f3d72c12d9.gif)

_running your first match_

## Get Useful Feedback. Quickly Iterate on Your Design.

Have an idea for a new strategy? Want to tweak your team ideas?

Before AI, to test a new team, you'd have to use your team for hours just to get a feel for how it performs. Now, you can get 100 matches of feedback in MINUTES. Use data, not feelings!

With one quick script, you can see how your bot fares against any number of opponent bots. Is your bot weak to Steel types? Special attacks? Stall teams? Trick Rooms? You'll know your bot's weaknesses and can improve your design as such.



## Love playing Pokemon but suck at programming?

Writing a Pokemon bot is a small, low-risk project where you can create something rewarding without a huge time investment. You don't have to create the next 'Deep Blue'!


## Already In Use In Classrooms
This software has been used for college-level game AI courses, where classes have run tournaments using students' submissions.


## Write in Any Language You Want

This project is mainly coded in Javascrpt / ES6, which I recommend for its ease of use. However, you can write code in any language you want.

Languages currently supported:
- jk nothing else yet


## Example Usage

### Install everything (Windows, Mac, Linux)
First, make sure you have installed [node and npm](https://docs.npmjs.com/getting-started/installing-node).
```bash
node -v
# 6.0.0 or greater plz!

# inside your bot's directory
npm install leftovers-again

# run generator script
node node_modules/leftovers-again/src/scripts/generate.js

# afterwards, your bot is ready to go
```bash
npm run server                          # in a separate window
npm install
npm start -- --opponent=randumb
```

### Run a `randombattle` between two bots:
```bash
npm run server                          # in a separate window
npm start -- stabby --opponent=randumb
```

### Play against your bot, via your web browser
```bash
npm run server                          # in a separate window
npm start
# browse to: http://play.pokemonshowdown.com/?~~localhost:8000
```

### Launch your bot on Cyberdyne
```bash
npm start -- --production
# browse to: http://cyberdyne.psim.us/
```


FAQ/Getting started guide [here](https://github.com/dramamine/leftovers-again/blob/master/FAQ.md).

Code documentation [here](http://metal-heart.org/static/docs/).

Contribution guide [here](https://github.com/dramamine/leftovers-again/blob/master/CONTRIBUTING.md).

Changelog [here](https://github.com/dramamine/leftovers-again/blob/master/CHANGELOG.md).


CHANGELOG.md
[0.11.5] - 2017-09-18
- Handle empty requests

[0.11.4] - 2017-08-31
- Hacky fix for nickname collisions

[0.11.3] - 2017-08-29
- Merge PR about toxic spikes stacking
- Change all formats to be gen7

[0.11.2] - 2017-08-27
- Update login command

[0.11.1] - 2017-08-24
- Update documentation

[0.11.0] - 2017-08-23
- Move from Babel to plain Node
- Log match starts better

[0.10.0] - 2016-11-27
- Update datasets for Gen7
- Implement Z-Moves

[0.9.8] - 2016-11-20
- Add the top two Brazillian bots

[0.9.6] - 2016-10-31
- Fixing gross issues with 'replace' and updating idents
- Really confident Zoroark & detailschange & formechange work
- Updated unit tests

[0.9.5] - 2016-10-30
- Switch module resolver
- Fix the 'develop' script to put files in the right directories
- Disconnect on opponent error
- Fix 'active mon during teamPreview' bug
- Update generate.js templates
- Fix watch commands for myself and for generate.js templates

[0.9.4] - 2016-10-27
- Change match-display to look a little nicer

[0.9.3] - 2016-10-27
- Fix issue with finding your Pokemon after forme/details change
- Clean up logging a bit

[0.9.0] - 2016-10-27
- Handle mega-evolutions and forme changes

[0.8.3] - 2016-10-25
- timer functionality (very hacky, not sure if its worth using)
- Forfeit on invalid decisions
- roundrobin.js script for running tourneys

[0.8.2] - 2016-10-16
- Forfeit matches if your bot throws an error
- Forfeit matches if you get a 'cant' message, and it 'cant' be helped
- Fully implement 'matches' param

[0.8.1] - 2016-10-12
- Fix for side effects not being properly written

[0.8.0] - 2016-10-02
- Tons of linting-related updates and npm script updates
- Better path handling for spawned opponents
- Integration test for node module

[0.7.1] - 2016-10-01
- Necessary fixes to templates; now you can load templated bots from places other than their root directory
- --results parameter, check results/results.csv by default

[0.7.0] - 2016-09-30
- Some stuff about the --interactive flag
- Implement --server flag
- Remove compiled lib files from git
- Major cleanup of challenger class

[0.6.4] - 2016-06-24
- Fixes for npm modules

[0.6.0] - 2016-06-20
- Make this work with npm
- `generate` script
- Windows compatibility

0.4.1 => 0.5.0
- Fix teams that break the single species rule (via nicknames)
- Old typechart method is deprecated, use Typechart.compare
- Cancel challenges sometimes
- Automatically mega-evolve
- Teampreview now shows opponent's Pokemons, as you'd expect

10/18/2015
npm run develop -- --bot=randombattle/stabby.js

10/9/2015
Special states:
- noCancel: this seems ok, you can't cancel yr moves.
- trapped: I think this means you can't switch out

10/4/2015
Really pulling off some shenanigans here. Currently only tested with 'randumb'. Tasks for next week:
- remove all 'state'ness from the battle class
- is everything migrated over??
  * include battleID in the state
- add validity checks (ex. don't send more than one active in a singles match)
- validate 'moves', what are they even doing right now?
- consider removing the 'state' variables from battlestore & pokemon
- unit tests for new classes.
- move 'challenge issuing' stuff into its own class maybe

Remember, you started rearranging everything because you were mad that you had to instantiate an AI to build a team, before you get any battle data. Are you solving that?
- consider creating a battle object, and setup listeners on the 'init' message? this is the first message with the battle ID.
- eh, maybe just say fuck it, make team() static
- lots of stuff is static, like all the metadata

9/11/2015
- Now two bots can fight each other to the end! In one terminal, run `babel-node scripts/spawn.js` (spawns martenbot by default) and run `babel-node src/main.js --bot=randombattle/stabby.js` in the other.


9/1/2015
- add a data store with battle info to connect.js or whatever. this might need to be an object with the battleId as its key to support multiple battles.
- when we get the 'turn' message (or something), call getMove() on your AI object.



API HACKING
example search:
http://replay.pokemonshowdown.com/search?user=DKFirelord


connect.js
const WebSocket = require('ws');
const querystring = require('querystring');
const http = require('http');
const https = require('https');
const url = require('url');
const ws = new WebSocket('ws://localhost:8000/showdown/websocket');

ws.on('open', function() {
  return ws.send('something');
});

ws.on('message', function(msg) {
  console.log('received: %s', msg);
  const messages = msg.split('\n');
  for(var i=0; i<messages.length; i++) {
    handleMessage(messages[i]);
  };
});

const ActionUrl = url.parse('https://play.pokemonshowdown.com/~~' + 'localhost:8000' + '/action.php');

const Config = {
  nick: '5nowden' + Math.floor(Math.random() * 10000),
  chatroom: 'lobby',
  battletype: 'randombattle'
};

const toId = function(text) {
  return text.toLowerCase().replace(/[^a-z0-9]/g, '');
};

const handleMessage = function(message) {
  var data, id, req, requestOptions, spl, str;
  spl = message.split('|');
  switch (spl[1]) {
    case 'updateuser':
      if (spl[2] !== Config.nick) {
        console.log('nickname was ', spl[2], ' expecting ', Config.nick);
        return;
      }
      if (spl[3] !== '1') {
        console.error('failed to log in, still guest');
        return;
      }
      console.log('logged in as ' + spl[2]);

      ws.send('|/join ' + Config.chatroom);
      break;
      // return console.log('would send', '|/blockchallenges');
    case 'challstr':
      console.info('received challstr, logging in...');
      id = spl[2];
      str = spl[3];
      requestOptions = {
        hostname: ActionUrl.hostname,
        port: ActionUrl.port,
        path: ActionUrl.pathname,
        agent: false
      };
      data = '';
      if (!Config.pass) {
        requestOptions.method = 'GET';
        requestOptions.path += '?act=getassertion&userid=' + toId(Config.nick) + '&challengekeyid=' + id + '&challenge=' + str;
      } else {
        requestOptions.method = 'POST';
        data = 'act=login&name=' + Config.nick + '&pass=' + Config.pass + '&challengekeyid=' + id + '&challenge=' + str;
        requestOptions.headers = {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Content-Length': data.length
        };
      }
      req = https.request(requestOptions, ( (res) => {
        res.setEncoding('utf8');
        data = '';
        res.on('data', (chunk) => {
          return data += chunk;
        });
        return res.on('end', ( () => {
          var e;
          if (data === ';') {
            console.error('failed to log in; nick is registered - invalid or no password given');
            process.exit(-1);
          }
          if (data.length < 50) {
            console.error('failed to log in: ' + data);
            process.exit(-1);
          }
          if (data.indexOf('heavy load') !== -1) {
            console.error('the login server is under heavy load; trying again in one minute');
            setTimeout((function() {
              return this.handleMessage(message);
            }).bind(this), 60 * 1000);
            return;
          }
          if (data.substr(0, 16) === '<!DOCTYPE html>') {
            console.error('Connection error 522; trying agian in one minute');
            setTimeout((function() {
              return this.handleMessage(message);
            }).bind(this), 60 * 1000);
            return;
          }
          try {
            data = JSON.parse(data.substr(1));
            if (data.actionsuccess) {
              data = data.assertion;
            } else {
              error('could not log in; action was not successful: ' + JSON.stringify(data));
              process.exit(-1);
            }
          } catch (_error) {
            e = _error;
            console.error('error trying to parse data:', e);
          }
          return ws.send('|/trn ' + Config.nick + ',0,' + data);
        }));
      }));

      req.on('error', (err) => {
        return console.error('login error: ' + err.stack);
      });

      if (data) {
        req.write(data);
      }
      return req.end();
    case 'j':
    case 'l':
    case ':':
    case 'init':
      break;
      // this is a multiline message.
      // message.split('\n');
      // break;
    case 'users':
      var userList = spl[2].split(', ');
      let user;
      // userlist[0] is just the count of users. skip it
      for(var i=1; i<userList.length; i++) {
        user = userList[i];
        // don't challenge yourself
        if(Config.user !== user) {
          console.log('challenging user..', userList[i]);
          ws.send('|/challenge ' + userList[i] + ', ' + Config.battletype);
        }
      }
      break;
    case 'updatechallenges':
      let challenges = JSON.parse(spl[2]);
      console.log(challenges);
      for (let username in challenges.challengesFrom) {
        // only accept battles of the type we're designed for
        if (challenges.challengesFrom[username] == Config.battletype) {
          // this is the point at which we need to pick a team!
          // TODO use promises here to maybe wait for user to pick a team
          // team message is: /utm ('use team')
          ws.send('|/accept ' + username);
        }
      }
    default:
      return console.log('not really handling message', spl[1]);
  }
};



function exitHandler(options, err) {
    if (options.cleanup) console.log('clean');
    if (err) console.log(err.stack);
    ws.close();
    if (options.exit) process.exit();
}

//do something when app is closing
process.on('exit', exitHandler.bind(null,{cleanup:true}));

//catches ctrl+c event
process.on('SIGINT', exitHandler.bind(null, {exit:true}));

//catches uncaught exceptions
process.on('uncaughtException', exitHandler.bind(null, {exit:true}));